presto-main/src/main/java/io/prestosql/operator/scalar/ArrayCombinationsFunction.java

```diff
@@ -0,0 +1,97 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.operator.scalar;
+
+import io.prestosql.spi.block.Block;
+import io.prestosql.spi.block.BlockBuilder;
+import io.prestosql.spi.block.PageBuilderStatus;
+import io.prestosql.spi.function.Description;
+import io.prestosql.spi.function.ScalarFunction;
+import io.prestosql.spi.function.SqlType;
+import io.prestosql.spi.function.TypeParameter;
+import io.prestosql.spi.type.ArrayType;
+import io.prestosql.spi.type.Type;
+
+import static io.prestosql.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static io.prestosql.spi.type.StandardTypes.INTEGER;
+import static io.prestosql.util.Failures.checkCondition;
+import static java.lang.StrictMath.toIntExact;
+import static java.util.Arrays.setAll;
+
+@ScalarFunction("combinations")
+@Description("Return k-element subsets from array")
+public final class ArrayCombinationsFunction
+{
+    private ArrayCombinationsFunction() {}
+
+    @TypeParameter("T")
+    @SqlType("array(array(T))")
+    public static Block combinations(
+            @TypeParameter("T") Type elementType,
+            @SqlType("array(T)") Block array,
+            @SqlType(INTEGER) long k)
+    {
+        int n = array.getPositionCount();
+        checkCondition(k >= 0, INVALID_FUNCTION_ARGUMENT, "k must be positive");
+
+        ArrayType arrayType = new ArrayType(elementType);
+        if (k > n) {
+            return arrayType.createBlockBuilder(new PageBuilderStatus().createBlockBuilderStatus(), 0).build();
+        }
+
+        BlockBuilder resultBuilder = arrayType.createBlockBuilder(new PageBuilderStatus().createBlockBuilderStatus(), 10_000 /* TODO n! / k! (n-k)! , but mind overflow */);
+        int[] combination = firstCombination(toIntExact(k));
+        do {
+            BlockBuilder arrayBuilder = elementType.createBlockBuilder(new PageBuilderStatus().createBlockBuilderStatus(), toIntExact(k));
+            for (int position : combination) {
+                elementType.appendTo(array, position, arrayBuilder);
+            }
+            arrayType.writeObject(resultBuilder, arrayBuilder.build());
+        }
+        while (nextCombination(combination, n));
+
+        return resultBuilder.build();
+    }
+
+    private static int[] firstCombination(int k)
+    {
+        int[] combination = new int[k];
+        setAll(combination, i -> i);
+        return combination;
+    }
+
+    private static boolean nextCombination(int[] combination, int n)
```

Comment:
There's a supposedly better algorithm to do this called "revolving door gray code". It's in Knuth's The Art of Computer Programming. Here's an implementation I wrote in Java (don't ask me why it works :) )

```java
private static <T> List<List<T>> combinations(List<T> list, int k)
{
    List<List<T>> result = new ArrayList<>();

    int[] combination = new int[k + 1];

    // initialize
    for (int i = 0; i < k; i++) {
        combination[i] = i;
    }
    combination[k] = list.size();

    while (true) {
        result.add(emit(list, combination, k));

        int i;
        if (k % 2 == 1) {
            if (combination[0] + 1 < combination[1]) {
                combination[0]++;
                continue;
            }

            i = 1;
        }
        else {
            if (combination[0] > 0) {
                combination[0]--;
                continue;
            }

            i = 1;

            if (increase(combination, i)) {
                continue;
            }
            i++;

            if (k == 2) {
                return result;
            }
        }

        while (true) {
            if (decrease(combination, i)) {
                break;
            }
            i++;

            if (increase(combination, i)) {
                break;
            }
            i++;

            if (i == k) {
                return result;
            }
        }
    }
}

private static boolean increase(int[] combination, int i)
{
    if (combination[i] + 1 < combination[i + 1]) {
        combination[i - 1] = combination[i];
        combination[i]++;
        return true;
    }
    return false;
}

private static boolean decrease(int[] combination, int i)
{
    if (combination[i] >= i + 1) {
        combination[i] = combination[i - 1];
        combination[i - 1] = i - 1;
        return true;
    }

    return false;
}

private static <T> List<T> emit(List<T> list, int[] combination, int k)
{
    List<T> result = new ArrayList<>(k);
    for (int i = 0; i < k; i++) {
        result.add(list.get(combination[i]));
    }
    return result;
}
```

Reply:
As discussed offline. For now i keep my (simplistic) code and we will revisit this when needed.
My assumption is that the function is useful only for small combination sizes.